### PR TITLE
feat(gemm): add SM100 dual-BF16 weight GEMM

### PR DESCRIPTION
**Pull request by xcandyyp**
### Motivation

This PR adds an SM100 dual-BF16 weight GEMM for BF16 activations and FP32 weights. It targets latency-sensitive FP32-weight workloads, especially MoE router GEMMs and sparse-attention state-compression GEMMs.

The FP32 weight is prepared once as two contiguous BF16 matrices:

```text
weight_high = BF16(weight_fp32)
weight_low  = BF16((weight_fp32 - FP32(weight_high)) * 256)
weight_fp32 ≈ FP32(weight_high) + FP32(weight_low) / 256
```

The runtime kernel evaluates both BF16 components with FP32 accumulation and supports either BF16 or FP32 output. Weight preparation is kept outside the timed GEMM path.

### What changed

- Added three CUTLASS/CuTe SM100 kernels and a device-aware dispatcher:
  - split-K 1SM for small-M problems;
  - persistent 1SM for intermediate-M shapes with an aligned N dimension;
  - cluster 2SM for large-M shapes and compatibility/tail cases.
- Added TVM-FFI JIT and AOT registration for exact SM100 (`sm_100a`).
- Added the public APIs:
  - `prepare_dual_bf16_weights()`;
  - `dual_bf16_weight_gemm_workspace_size()`;
  - `mm_bf16_dual_weight()`.
- Added caller-owned workspace support. Only split-K requires workspace (FP32 partial outputs plus int32 completion counters); the persistent 1SM and cluster 2SM paths require zero bytes.
- Added per-device/per-stream cached workspace as a convenience fallback. CUDA Graph and concurrent multi-stream callers can provide explicit independent buffers.
- Added BF16/FP32 output handling to the FlashInfer trace template and generated trace fixture.
- Added API documentation and a benchmark against a strict FP32 PyTorch/cuBLAS baseline.

### Dispatch policy and kernel configuration

All paths require positive `M`, `N`, and `K`, exact SM100 (compute capability 10.0), and `K % 128 == 0`.

| Selected path | Dispatch condition | Tile/config selection | Workspace |
|---|---|---|---|
| split-K 1SM | `M <= 256` and `K >= 256` | `N_tile=64`, `K_tile=128`; `M_tile=8` for `M < 64`, otherwise `M_tile=16`; 4 stages. Split factor is selected from `{2, 4, 8}`: choose the largest valid factor whose task grid does not exceed the SM count, or the smallest valid factor if all valid grids exceed it. | FP32 partial outputs plus one int32 counter per output tile |
| persistent 1SM | Not split-K, `M < 1024`, and `N % 64 == 0` | `N_tile=64`, `M_tile=16`, `K_tile=128`, 6 stages; persistent grid capped by the SM count | None |
| cluster 2SM | All remaining valid shapes | 2-CTA clusters, `K_tile=128`. If `N <= 128`: `N_tile=128`, with `(M_tile, stages)=(16,6)` for `M <= 1024`, `(32,6)` for `1024 < M < 4096`, and `(64,5)` for `M >= 4096`. If `N > 128`: `N_tile=256`, with `M_tile=16/32/64` at the same M boundaries and 3 stages. | None |

`K=128` intentionally bypasses split-K and is dispatched to the persistent 1SM or cluster 2SM path. For `M < 1024`, an N tail (`N % 64 != 0`) uses the cluster 2SM compatibility path.

### Correctness and engineering coverage

- Boundary correctness includes the minimum legal `K=128`, minimum dimensions, the `M=256/257`, `M=1023/1024`, and `M=4095/4096` transitions, the 1SM `N % 64` requirement, the 2SM `N=128/129` switch, and both BF16 and FP32 outputs.
- Added explicit workspace size/alignment/reuse validation.
- Added CUDA Graph capture/replay coverage for all three dispatch paths.
- Added concurrent multi-stream coverage for all three dispatch paths.
- Added an AOT registration test and trace output-dtype tests.
- The C++ operator performs no internal device allocation.

### Validation performed

- `48` targeted GEMM/JIT/trace/benchmark tests passed on B200 (`41` operator tests plus `7` JIT/trace/benchmark integration checks).
- `pre-commit run --all-files` passed.
- Compute Sanitizer covered all three dispatch paths:
  - `memcheck`: 0 errors;
  - `synccheck`: 0 errors;
  - `racecheck`: 0 errors after applying the included narrow suppression for the Compute Sanitizer 2026.1 false positive in CUTLASS `Allocator2Sm::allocate` (`tcgen05.alloc.cta_group::2`). All other accesses remain checked.

The sanitizer driver and commands are documented in `tests/gemm/dual_bf16_weight_gemm_sanitizer.py`. The suppression should be re-tested when Compute Sanitizer is upgraded.

### Performance

Environment: NVIDIA B200 (SM100), driver 595.71.05, CUDA 13.2, PyTorch 2.11.0a0. Timings use CUDA events, cold-L2-cache measurement, 10 warm-up iterations, and 50 measured iterations. The output is FP32, `N=192`, `K=4096`, TF32 is disabled, and PyTorch is forced to use cuBLAS. Weight preparation and BF16-to-FP32 activation conversion are excluded from timing.

| Dispatch path | Shape `(M, N, K)` | dual-BF16 | FP32 cuBLAS | Speedup |
|---|---:|---:|---:|---:|
| split-K 1SM | `(128, 192, 4096)` | 0.011 ms | 0.046 ms | 4.361× |
| persistent 1SM | `(512, 192, 4096)` | 0.014 ms | 0.073 ms | 5.422× |
| cluster 2SM | `(1024, 192, 4096)` | 0.018 ms | 0.109 ms | 5.911× |

The dual-BF16 throughput calculation counts two BF16 GEMMs (`4MNK` operations), while the FP32 cuBLAS baseline counts one FP32 GEMM (`2MNK`). The speedup is the directly measured latency ratio `cuBLAS / dual-BF16`. The dual-BF16 result also passed the benchmark's accuracy check against the FP32 cuBLAS result (cosine similarity 1.00000000 in the reported run).

### Current scope

- Exact SM100 only; the implementation uses SM100a-specific TMA/UMMA features.
- Inputs are contiguous BF16 activation/high/low tensors on the same CUDA device.
- The reduction dimension must be a positive multiple of 128.
- Output dtype is BF16 or FP32.

## 🔍 Related Issues

N/A

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used my preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] Relevant local tests are passing; the full architecture matrix will run in CI.

## Reviewer Notes

Please focus on the three-path dispatch thresholds, tile/stage selection, split-K workspace lifecycle, and the 2SM cluster barrier/TMEM lifecycle. The included racecheck suppression is intentionally restricted to the known Compute Sanitizer 2026.1 report in CUTLASS `Allocator2Sm::allocate` and should not mask other kernel accesses.

## Summary by CodeRabbit

- **New Features**
  - Added SM100/Blackwell support for GEMM with dual-BF16 weights.
  - Added utilities to prepare dual-BF16 weights and calculate required workspace.
  - Supports BF16 and FP32 outputs, preallocated outputs, workspace reuse, and CUDA graph capture.
  - Added automatic selection across split-K, single-SM, and clustered execution paths.
  - Added tracing, benchmarking, and ahead-of-time build support.

- **Documentation**
  - Added API documentation and benchmark guidance.

- **Tests**
  - Added coverage for accuracy, validation, dispatch, tracing, sanitizer checks, and workspace reuse.

### CODE REVIEW

**Comment by xcandyyp** [comment]
Hi maintainers, I’ve addressed all current review findings in commit 73b0ef23. Local pre-commit, targeted B200 tests, documentation checks, and Compute Sanitizer checks for all three dispatch paths pass. Could someone with CI permission please trigger the public CI on the latest commit? Thank you.

**Comment by xcandyyp** [comment]
@yongwww The branch has now been rebased onto the latest `main` (base `6d5ef77f`; PR head `5d420bb1`). Post-rebase validation passed:

- full pre-commit
- 48 targeted GEMM/JIT/trace/benchmark tests on B200
- Compute Sanitizer memcheck, racecheck, and synccheck across all three dispatch paths
- refreshed FP32 cuBLAS comparisons for all three dispatch paths

Could you or another maintainer with CI permission please trigger the public CI for this latest commit? Thank you.

File: flashinfer/gemm/gemm_dual_bf16_weight.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+    normalized = _normalize_cuda_device(device)
+    _require_exact_sm100(normalized)
+    module = _get_dual_bf16_weight_gemm_module()
+    return int(module.kernel_kind(m, n, k, get_device_index(normalized)))
+
+
+@flashinfer_api
+@torch.no_grad()
+def prepare_dual_bf16_weights(
```
[the remark this answers is not in the document]
Comment (xcandyyp) [reply flashinfer/gemm/gemm_dual_bf16_weight.py:123]: Fixed in `5d420bb1`: `prepare_dual_bf16_weights` now documents `Args`, `Returns`, and `Raises`, and the post-rebase PR documentation check reports no new findings.

**Comment by xcandyyp** [comment]
@dhiraj113 Hi Dhiraj, would you have time to review this SM100 dual-BF16 weight GEMM PR? Given your recent reviews of Blackwell and GEMM-related changes, your feedback on the three-path dispatch, split-K synchronization/workspace handling, and SM100 integration would be especially valuable.

The branch has been rebased onto the latest `main`:
- base: `fbad18cb`
- head: `18ca5869`
- status: 0 commits behind / 2 commits ahead

Validation on NVIDIA B200:
- `pre-commit run --all-files`: passed
- 48 targeted GEMM/JIT/trace/benchmark tests: passed
- Compute Sanitizer 2026.1 `memcheck`, `racecheck`, and `synccheck`: 0 errors across split-K 1SM, persistent 1SM, and cluster 2SM paths

The PR Overview has also been refreshed with the latest benchmark measurements. Public CI for this new head still requires an authorized maintainer trigger; if the change looks suitable for CI, could you please help trigger it? Thank you!
